File: week1/lcm.js
const { performance } = require("perf_hooks");

function lcm(a, b) {
  //check for either a or b being a multiple
  if (a % b === 0) {
    return b;
  }
  if (b % a === 0) {
    return a;
  }
  let v = 1;
  //count starts at the smaller of the two nums
  let c = a < b ? a : b;
  let r = false;
  const aMults = [a];
  const bMults = [b];
  const check = (l, lArr, jArr) => {
    if (jArr.includes(l * c)) {
      v = l * c;
      r = true;
    } else {
      lArr.push(l * c);
    }
  };

  while (!r && (a < b ? a : b)) {
    check(b, bMults, aMults);
    check(a, aMults, bMults);
    c++;
  }
  return v;
}

const perfTest = pair => {
  const start = performance.now();
  const val = lcm(...pair);
  const end = performance.now();
  console.log(`lcm(${pair}) returns ${val} in ${end - start} ms`);
};

const generateTests = () => {
  for (let i = 0; i < 10; i++) {
    let pair = [];
    for (let i = 0; i < 2; i++) {
      pair.push(Math.floor(Math.random() * Math.floor(30)));
    }
    console.log("testing pair ", pair);
    perfTest(pair);
  }
};

generateTests();

File: week1/lcm.py
import time
import random


def lcm(a, b):
    if a % b == 0:
        return b
    if b % a == 0:
        return a
    v = 1

    if a < b:
        c = a
    else:
        c = b

    r = False

    mults = {"aMults": [a], "bMults": [b]}

    while r == False:
        if a * c in mults["bMults"]:
            v = a * c
            r = True
        else:
            mults["aMults"].append(a * c)
        if b * c in mults["aMults"]:
            v = b * c
            r = True
        else:
            mults["bMults"].append(b * c)
        c = c + 1
    return v


def perfTest(pair):
    print("pair is ", pair)
    a, b = pair
    start = time.time()
    val = lcm(a, b)
    end = time.time()
    print(val)
    print(format(end - start, "f"))


def generateTests():
    i = 10
    while i > 0:
        c = random.randint(1, 30)
        d = random.randint(1, 30)
        perfTest([c, d])
        i = i - 1


generateTests()
